File: src/main/scala/day8.scala
object day8 extends App {

  val data = Util.getDataInLines("day8").map(x => {
    val instruction = x.split(" ").head
    val argument = x.split(" ").last.toInt
    new Instruction(instruction, argument)
  })

  @scala.annotation.tailrec
  def compute(data: List[Instruction], i: Int, acc: Int, alreadyVisitedInstructions: Set[Int]): Int = {
    if (alreadyVisitedInstructions.contains(i)) {
      println("loop detected!")
      return acc
    }
    val instruction = data(i)
    if (instruction.instruction == "nop") {
      compute(data, i + 1, acc, alreadyVisitedInstructions.incl(i))
    } else if (instruction.instruction == "jmp") {
      compute(data, i + instruction.argument, acc, alreadyVisitedInstructions.incl(i))
    } else if (instruction.instruction == "acc") {
      compute(data, i + 1, acc + instruction.argument, alreadyVisitedInstructions.incl(i))
    } else {
      0
    }
  }

  val result = compute(data, 0, 0, Set.empty)
  println(result)

}

class Instruction(val instruction: String, val argument: Int) {
  override def toString: String = s"instruction $instruction, argument: $argument"
}

File: src/main/ruby/day8.rb
require 'set'

file = File.open("../resources/inputs/day8_modified.txt")
data = file.readlines
file.close

i = 0
acc = 0
already_visited_instructions = Set.new

def compute(s, i, acc)
  operation, argument = s.split(' ')
  puts "i: #{i}, acc: #{acc}, operation: #{operation}, argument: #{argument}"
  case operation
  when 'nop'
    [i + 1, acc]
  when 'acc'
    [i + 1, acc + argument.to_i]
  when 'jmp'
    [i + argument.to_i, acc]
  end
end

loop do
  if already_visited_instructions.include? i
    puts "loop detected!"
    break acc
  end
  if i == data.size
    puts 'made it to the end!'
    break acc
  end

  already_visited_instructions.add i
  i, acc = compute(data[i], i, acc)
end

puts acc

